File: 206.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        if head == None or head.next == None:
            return head
        l = head
        r = head.next
        while r != None:
            tmp = r.next
            r.next = l
            l = r
            r = tmp
        head.next = None
        return l

File: 206.cc
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */

class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        if (head == NULL) return NULL;
        ListNode* res = head;
        while (res->next != NULL) res = res->next;
        ListNode *tail = revList(head);
        tail->next = NULL;
        return res;
    }
    
    ListNode* revList(ListNode* head) {
        if (head->next == NULL) return head;
        ListNode* tail = revList(head->next);
        tail->next = head;
        return head;
    }
}; 

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        ListNode* newhead = NULL;
        while (head != NULL) {
            ListNode* tmp = newhead;
            ListNode* tmp2 = head->next;
            newhead = head;
            head->next = tmp;
            head = tmp2;
        }
        return newhead;
    }
};